clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -93,14 +93,11 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
 	}
 
 	glog.Info("Creating external cluster")
-	externalClient, err := d.createExternalCluster()
+	externalClient, cleanupExternalCluster, err := d.createExternalCluster()
+	defer cleanupExternalCluster()
```

Comment:
Is it always valid to call the `defer` if there's an error?

Reply:
looking at the change, the cleanupFn is a noop if it fails to create.
it is only set to do cleanup if it manages to create successfully in createExternalCluster.

Reply:
My concern would be ongoing maintenance holds that invariant to be true.

Reply:
The previous logic we had (where the defer only happens after the error) was confusing. Enough that there was a PR out to fix the "bug". I figure if there is only one location that needs to determine what cleanup to do (if any) is going to be easier to understand and maintain.

In terms of holding invariant, the unit tests make sure things get cleaned up when expected to. If we are calling a invalid defer, it will break that test. https://github.com/kubernetes-sigs/cluster-api/blob/master/clusterctl/clusterdeployer/clusterdeployer_test.go#L140